cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -175,3 +175,40 @@ def dephase(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
     return transformer_primitives.map_operations(
         circuit, dephase, deep=True, tags_to_ignore=ignored
     ).unfreeze()
+
+
+@transformer_api.transformer
+def drop_terminal_measurements(
+    circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+) -> 'cirq.Circuit':
+    """Removes terminal measurements from a circuit.
+
+    This transformer is helpful when trying to capture the final state vector
+    of a circuit with many terminal measurements, as simulating the circuit
+    with those measurements in place would otherwise collapse the final state.
+
+    Args:
+        circuit: The circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options
+            for transformers.
+    Returns:
+        A copy of the circuit, with identity or X gates in place of terminal
+        measurements.
+    Raises:
+        ValueError: if the circuit contains non-terminal measurements.
+    """
+
+    def flip_inversion(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if isinstance(op.gate, ops.MeasurementGate):
```

Comment:
Can / Should we use a combination of `protocols.is_measurement` + `protocols.decompose` to also support `PauliMeasurementGate` ? 

Note that `circuit.are_all_measurements_terminal` also checks for a measurement gate using `protocols.is_measurement`.

Reply:
Note that `protocols.is_measurement` returns True for a CircuitOperation that contains a measurement - `circuit.are_all_measurements_terminal` can use it because it checks the subcircuit case, but here I think we're stuck checking types directly.

RE `PauliMeasurementGate`: I think either we treat it as its components, in which case it's not a terminal measurement (since the result is e.g. XMX), or we treat it as an alternate-basis measurement, in which case it converts to `cirq.I` since it has no invert mask. Which do you think is preferable?

Reply:
You are right.  Since `PauliMeasurementGate` decomposes into a non-terminal measurement, we can leave this implementation as it is because even after expanding the decompose, the measurement will not be a terminal measurement.